content/academy/puppeteer_playwright/collecting_data.md

```diff
@@ -0,0 +1,284 @@
+
+## Setup
+
+Here is the base set up for our code, upon which we'll be building off of in this lesson:
+
+```marked-tabs
+<marked-tab header="Playwright" lang="javascript">
+import { chromium } from 'playwright';
+
+const browser = await chromium.launch({ headless: false });
+const page = await browser.newPage();
+
+await page.goto('https://demo-webstore.apify.org/search/on-sale');
+
+await page.waitForLoadState('networkidle');
+
+// code will go here
+
+await page.waitForTimeout(10000)
+
+await browser.close();
+</marked-tab>
+<marked-tab header="Puppeteer" lang="javascript">
+import puppeteer from 'puppeteer';
+
+const browser = await puppeteer.launch({ headless: false });
+const page = await browser.newPage();
+
+await page.goto('https://demo-webstore.apify.org/search/on-sale');
+
+await page.waitForNetworkIdle();
+
+// code will go here
+
+await page.waitForTimeout(10000)
+
+await browser.close();
+</marked-tab>
+```
+
+> Notice the slightly different syntax between Playwright and Puppeteer with `waitForLoadState('networkidle')` and `waitForNetworkIdle()`. Sometimes, the differences are fairly subtle like this, but later on we'll run into some more significant differences.
+
+## [](#collecting-in-page-evaluate) Collecting in the browser context
+
+Whatever is returned from the callback function in `page.evaluate()` will be returned by the evaluate function, which means that we can set it to a variable like so:
+
+```JavaScript
+const products = await page.evaluate(() => {});
+```
+
+We'll be returning a bunch of product objects from this function, which will be accessible back in our Node.js context after the promise has resolved. Let's now go ahead and write some data collection code to collect each product:
+
+```JavaScript
+const products = await page.evaluate(() => {
+    return [...document.querySelectorAll('a[class*="ProductCard_root"]')].map((elem) => {
+        const name = elem.querySelector('h3[class*="ProductCard_name"]').textContent;
+        const price = elem.querySelector('div[class*="ProductCard_price"]').textContent;
+
+        return {
+            name,
+            price,
+        };
+    });
+});
+
+console.log(products);
+```
+
+When we run this code, we see this logged to our console:
+
+![Products logged to the console]({{@asset puppeteer_playwright/images/log-products.webp}})
+
+## [](#using-jquery) Using jQuery
+
+With the `page.addScriptTag()` function and the latest [jQuery CDN link](https://releases.jquery.com/), we can inject jQuery into the current page:
+
+```JavaScript
+await page.addScriptTag({ url: 'https://code.jquery.com/jquery-3.6.0.min.js' });
+```
+
+Now, since we're able to use jQuery, let's translate our vanilla JavaScript code within the `page.evaluate()` function to jQuery:
+
+```JavaScript
+await page.addScriptTag({ url: 'https://code.jquery.com/jquery-3.6.0.min.js' });
+
+const products = await page.evaluate(() => {
+    return [...$('a[class*="ProductCard_root"]')].map((elem) => {
```

Comment:
This is more readable thanks to being shorter, but I think jQuery has a native method that returns an array, which I think would be more readable than the `return [...spread].map(). A lot of things happening on one line.